pmlc/dialect/tile/transforms/pad_ranges.cc

```diff
@@ -0,0 +1,148 @@
+#include "pmlc/dialect/tile/transforms/pass_detail.h"
+#include "pmlc/util/ident.h"
+#include "pmlc/util/logging.h"
+#include "pmlc/util/strides.h"
+
+namespace pmlc::dialect::tile {
+
+using namespace mlir; // NOLINT
+
+namespace {
+
+struct PadRangesPass : public PadRangesBase<PadRangesPass> {
+  int64_t maxPowerOfTwo = 16;
+  int64_t minPowerOfTwo = 8;
+  int64_t maxIncreasePercent = 5;
+  void runOnFunction() final;
+};
+
+void PadRangesPass::runOnFunction() {
+  auto func = getFunction();
+  func.walk([&](ContractionOp op) {
+    // Skip some cases where the padding pass can't operate.
+    if (op.getNumSymbols()) {
+      op.emitRemark("padding cannot be run on symbolic contractions");
+      return;
+    }
+    if (!op.lowerBounds() || !op.upperBounds()) {
+      op.emitRemark("contraction bounds must be computed");
+      return;
+    }
+    // Extract which dimensions of the contraction are which output dimension
+    SmallVector<unsigned, 4> outIdx;
+    unsigned outRank = op.getResultType().getRank();
+    for (unsigned i = 0; i < outRank; i++) {
+      auto outDim = op.sink().getResult(i).dyn_cast<AffineDimExpr>();
+      if (!outDim) {
+        op.emitRemark("Invalid output access");
+        return;
+      }
+      outIdx[i] = outDim.getPosition();
+    }
+    // Verify bounds are simple
+    unsigned indexCount = op.lowerBounds()->getNumResults();
+    SmallVector<int64_t, 6> ranges;
+    for (unsigned i = 0; i < indexCount; i++) {
+      auto lowerConst =
+          op.lowerBounds()->getResult(i).dyn_cast<AffineConstantExpr>();
+      auto upperConst =
+          op.upperBounds()->getResult(i).dyn_cast<AffineConstantExpr>();
+      if (!lowerConst || lowerConst.getValue() != 0) {
+        op.emitRemark("Invalid lower bound");
+        return;
+      }
+      if (!upperConst) {
+        op.emitRemark("Invalid lower bound");
+        return;
+      }
+      ranges.push_back(upperConst.getValue() + 1);
+    }
+    // Get strides for output
+    auto outStrides = util::computeStrideArray(op.getResultType(), op.sink());
+    IVLOG(1, "outStrides = " << *outStrides);
+    // Decide wthat to round up
+    bool didChange = false;
+    SmallVector<int64_t, 6> newRanges = ranges;
+    for (unsigned i = 0; i < indexCount; i++) {
+      if (outStrides->strides[i] == 0) {
+        continue;
+      }
+      for (int64_t po2 = maxPowerOfTwo; po2 >= minPowerOfTwo; po2 /= 2) {
+        int64_t roundedRange = llvm::divideCeil(ranges[i], po2) * po2;
+        if (roundedRange == ranges[i]) {
+          break;
+        }
+        int64_t increase = roundedRange - ranges[i];
+        int64_t precIncrease = increase * 100 / ranges[i];
+        IVLOG(1, "Orig: " << ranges[i] << ", rounded: " << roundedRange
+                          << ", precIncrease = " << precIncrease);
+        if (precIncrease < maxIncreasePercent) {
+          newRanges[i] = roundedRange;
+          didChange = true;
+          break;
```

Comment:
This `break` implies that the cost of being closer to the "max" is always lower than being closer to the "min".  Do we know that will always be the case?

Reply:
Yes, in that if we can round to 16 for less than 5% and we can round to 8 for less than 5%, I would argue we should always round to 16.  I think that is a reasonable assumption.